### ISSUE TITLE
Improve error reporting for "evaluate constant expression" in TLC

### ISSUE THREAD

**Issue by manbansa** [tlaplus/tlaplus#179]
While trying to verify the definition of following operator using "evaluate constant expression", I am seeing some wired output.

Input:- 
`{s \in SUBSET {1,2,3,4} : \A x \in s : IF x = Max(s)                                                                    									THEN TRUE                                                                                        									ELSE (x+1) \in s}` 

Output:-
`{s \in {{}, {1}, {2}, {3}, {4}, {1, 2}, {1, 3}, {1, 4}, {2, 3}, {2, 4}, {3, 4}, {1, 2, 3}, {1, 2, 4}, {1, 3, 4}, {2, 3, 4}, {1, 2, 3, 4}} : <expression line 6, col 27 to line 6, col 376 of module MC> }`

Basically my definition of Max(s) operator was flawed 
`Max(s) == { x \in s : \A y \in s : x >= y}`

I took me a lot of time to realize  that Max(s) operator's definition is the culprit here. A better error reporting from TLC would have been really useful. Can you please improve the error reporting in TLC for "evaluate constant expression"?

**Comment by lemmy** [comment]
@inieves Do you have time to look into this issue?

**Comment by inieves** [comment]
@manbansa is the definition of Max(s) given above the flawed or unflawed version?  Can you tell me what both versions of Max(s) look like, and which is flawed vs. unflawed?

I have some ideas of how to improve the output, but I am curious to know if you have any particular ideas in mind.

**Comment by lemmy** [comment]
@inieves The def of `Max(s)` above returns a set whose element is the maximum element. The equality in IF THEN ELSE compares {max} to an element max.

**Comment by inieves** [comment]
@lemmy I can see that now, thanks for highlighting.  This would seem to be a type mismatch.  yes/no?  Is this type mismatch the source of the flaw that @manbansa is talking about?

**Comment by lemmy** [comment]
@inieves TLA+ is untyped: [It's sets all the way down](https://en.wikipedia.org/wiki/Zermelo%E2%80%93Fraenkel_set_theory)¹ which makes the above what Leslie calls a _silly expression_ (Specifying Systems page 67). TLC is no friend of silly expressions.

¹ The set of all sets is not a set (see [Russel's Paradox](https://en.wikipedia.org/wiki/Russell%27s_paradox))

**Comment by inieves** [comment]
@lemmy @manbansa I am not able to resolve these questions:
1) what is the correct implementation of Max(s)
2) what is the incorrect implementation of Max(s)
3) what is "better" error reporting in the context of the example above?

@lemmy I understand that TLA+ is typeless, an TLC does not like silly expressions.  Is strange output (above) generated because the expression is silly?  Would "better" error reporting be: "inform the user of the existence (and location) of a silly expression"?

**Comment by lemmy** [comment]
1. `Max(S) == CHOOSE s \in S: \A t \in (S \ {s}): t < s`
2. `Max(s) == { x \in s : \A y \in s : x >= y}`
3. TLC and the Toolbox correctly correctly an error with a stack if `{s \in SUBSET {1,2,3,4} : \A x \in s : IF x = Max(s) THEN TRUE ELSE (x+1) \in s}` gets evaluated as part of the regular init and next-state relation (ordinary model checking). It only produces this funny output when the expression is evaluated with "Evaluate Constant Expressions".

**Comment by inieves** [comment]
@lemmy this is helpful, thank you

**Comment by manbansa** [comment]
Hi inieves@
As Markus mentioned, the definition of Max(S) mentioned in the bug description is the flawed one.

**Comment by inieves** [comment]
Hi @lemmy ... i have finally gotten some more time to look into this, and made all necessary updates to my dev environment to be able to experiment and test...

The request for improved error reporting for evaluating constant expressions means that (in my opinion) we want to show a stack trace if/when an error occurs within evaluation of a constant expression.

The evaluation of constant definitions happens (at low level) in SpecProcessor.java at processConstantDefns @ 198.  There are a few calls to spec.eval(...) in there.  In one possible solution, the overall idea would be to make sure the call stack is on when those eval calls are made.  And then the call stack could be printed out upon termination.

It is not too hard to add the basic code to make the test case of @manbansa work... But I am concerned that extensive testing would be needed to ensure that other things didn't break.

I'd be happy to walk you through what I know about this, and how I would fix it, but I can't do this on a small time budget alone.

At least though now I have a pretty clear idea of where the problem is and how it might be fixed.  I think with your combined knowledge, a good solution is possible.

**Comment by lemmy** [comment]
@inieves Can you attach a patch to fix manbansa's scenario.  Running it on the test suite might help us better estimate the risk.

**Comment by inieves** [comment]
@lemmy I can see if a simple fix (patch) is possible.  My initial thought was that there was not a simple fix, and that some discussion would make sense before making larger changes.  If I can’t make a simple fix/patch, I will let you know.

**Comment by inieves** [comment]
@lemmy I can see that manbansa's original definition of Max(s) did not have the effect that he intended, but it seems to be valid TLA+ in the sense that no exceptions are thrown during parsing.  Yes it is a silly expression, but does that sillyness alone make it invalid TLA+ (and thus an error)?

I have found that in the file SemanticNode.java there is mention of a System Property (in the method toString() at line 291) called:
tla2sany.semantic.SemanticNode.showPlainFormulae


When I simulate the usage of this system property, I get this as the "value" in Toolbox:
{s \in {{}, {1}, {2}, {3}, {4}, {1, 2}, {1, 3}, {1, 4}, {2, 3}, {2, 4}, {3, 4}, {1, 2, 3}, {1, 2, 4}, {1, 3, 4}, {2, 3, 4}, {1, 2, 3, 4}} : <expression \Ax\ins:IFx=Max(s)THENTRUEELSE(x+1)\ins> }

And I get this in "user output" of Toolbox:
<<"$!@$!@$!@$!@$!", {s \in {{}, {1}, {2}, {3}, {4}, {1, 2}, {1, 3}, {1, 4}, {2, 3}, {2, 4}, {3, 4}, {1, 2, 3}, {1, 2, 4}, {1, 3, 4}, {2, 3, 4}, {1, 2, 3, 4}} : <expression \Ax\ins:IFx=Max(s)THENTRUEELSE(x+1)\ins> }>>


@manbansa would this additional error output have been helpful?  At least then you would not have had to refer back to the MC.tla file (which you may not even know exist[ed])

**Comment by lemmy** [comment]
@inieves I think what we really want is ECE to report the same error that ordinary model checking produces.

**Comment by inieves** [comment]
![Screen Shot 2019-05-22 at 5 21 05 PM](https://user-images.githubusercontent.com/656250/58353835-6b8f8100-7e24-11e9-8293-fbf19371b9ac.png)

@lemmy I have attached a screenshot of the error that ordinary model checking produces.  If you look closely, you can see that an error occurs because fingerprinting fails.  But in the case of evaluation of constant definitions/expressions, there is no fingerprinting, thus there is no fingerprint failure...

In the case of evaluation of constant definitions/expressions, TLA does not produce an error, it seems to ignore the fact that there is a comparison between integer and non-integer, and it produces actual output (the output mentioned in the very first message in this thread).

If you wan't the evaluation of constant definitions/expressions to produce the same error output as ordinary model checking, it seems that some variation of fingerprinting (at least) has to be added.  I think this goes far beyond adding a call stack or additional error output... this is an architectural change.

@lemmy my biggest questions at this point is:
1) should TLA be producing an error of any kind (even just fingerprint error) when doing ordinary model checking on manbansa's test case, or should it spit out similar output to the mode of evaluation of constant expressions? (is manbansa's original input valid TLA?)
2) is it correct that fingerprinting is failing on manbansa's input during ordinary model checking, or is that a bug?
3) To me, this difference in functionality between ordinary model checking and evaluation of constant expressions that manbansa has found is not just a difference in error reporting... it seems to be a fundamental difference in functionality... which is surprising, because I was thinking that evaluation of constant expressions was meant to be a way to experiment quickly with TLA as a calculator... but if they are not calculating exactly the same, that itself seems to be a bug.

**Comment by inieves** [comment]
hi @lemmy the simplest patch is to change SetPredValue.java:305-311 from:

`
try {
        if (TLCGlobals.expand) {
          Value  val = this.toSetEnum();
          return val.toString(sb, offset);
        }
      }
      catch (Throwable e) { /*SKIP*/ }
`

to just:

`
      if (TLCGlobals.expand) {
          Value  val = this.toSetEnum();
          return val.toString(sb, offset);
        }
`

There was an exception being silently caught here.  This will then allow a more reasonable error message to be produced:

==========

The `Evaluate Constant Expression� section�s evaluation failed.

Attempted to apply the operator overridden by the Java method
public static tlc2.value.impl.Value tlc2.module.TLC.PrintT(tlc2.value.impl.Value),
but it produced the following error:
Attempted to check equality of integer 1 with non-integer:
{1}

==========

The good news is that that message would probably have been the key to help manbansa.

To make the error messages for evaluating constant expressions fully up to par with normal model checking would require implementation of a new type of callstack and as well as other modifications to support setting the source (SemanticNode).

**Comment by lemmy** [comment]
Thanks @inieves for spotting the silently discarded exceptions.
